File: simplified-books-core/src/main/java/org/nypl/simplified/books/core/FeedEntryOPDS.java
package org.nypl.simplified.books.core;

import com.io7m.jnull.NullCheck;
import org.nypl.simplified.opds.core.OPDSAcquisitionFeedEntry;

/**
 * An entry from an OPDS feed.
 */

public final class FeedEntryOPDS implements FeedEntryType
{
  private static final long serialVersionUID = 1L;
  private final BookID                   book_id;
  private final OPDSAcquisitionFeedEntry entry;

  private FeedEntryOPDS(
    final BookID in_book_id,
    final OPDSAcquisitionFeedEntry in_e)
  {
    this.book_id = NullCheck.notNull(in_book_id);
    this.entry = NullCheck.notNull(in_e);
  }

  /**
   * Construct a feed entry from the given OPDS feed entry.
   *
   * @param e The entry
   *
   * @return A feed entry
   */

  public static FeedEntryType fromOPDSAcquisitionFeedEntry(
    final OPDSAcquisitionFeedEntry e)
  {
    return new FeedEntryOPDS(BookID.newIDFromEntry(e), e);
  }

  /**
   * @return The actual feed entry
   */

  public OPDSAcquisitionFeedEntry getFeedEntry()
  {
    return this.entry;
  }

  @Override public BookID getBookID()
  {
    return this.book_id;
  }

  @Override public <A, E extends Exception> A matchFeedEntry(
    final FeedEntryMatcherType<A, E> m)
    throws E
  {
    return m.onFeedEntryOPDS(this);
  }
}

File: simplified-books-core/src/main/java/org/nypl/simplified/books/core/FeedEntryOPDS.kt
package org.nypl.simplified.books.core

import com.io7m.jfunctional.OptionType

import org.nypl.simplified.opds.core.OPDSAcquisitionFeedEntry

import org.nypl.simplified.books.core.BookFormats.BookFormatDefinition

/**
 * An entry from an OPDS feed.
 */

data class FeedEntryOPDS private constructor(
  val actualBookID: BookID,
  val feedEntry: OPDSAcquisitionFeedEntry,
  val probableFormat: OptionType<BookFormatDefinition>) : FeedEntryType {

  override fun getBookID(): BookID {
    return this.actualBookID
  }

  override fun <A, E : Exception> matchFeedEntry(
    m: FeedEntryMatcherType<A, E>): A {
    return m.onFeedEntryOPDS(this)
  }

  companion object {

    private const val serialVersionUID = 1L

    /**
     * Construct a feed entry from the given OPDS feed entry.
     *
     * @param entry The entry
     * @return A feed entry
     */

    fun fromOPDSAcquisitionFeedEntry(entry: OPDSAcquisitionFeedEntry): FeedEntryType {
      return FeedEntryOPDS(BookID.newIDFromEntry(entry), entry, BookFormats.inferFormat(entry))
    }
  }
}
